src/main/java/com/palantir/gradle/versions/GradleWorkarounds.java

```diff
@@ -124,5 +130,82 @@ static boolean isFailOnVersionConflict(Configuration conf) {
         return conflictResolution == org.gradle.api.internal.artifacts.configurations.ConflictResolution.strict;
     }
 
+    @SuppressWarnings("CyclomaticComplexity")
+    public static void makeEvaluationDependOnSubprojectsToBeEvaluated(Project rootProject) {
+        if (!rootProject.getGradle().getStartParameter().isConfigureOnDemand()
+                || rootProject.getGradle().getStartParameter().isWriteDependencyLocks()
+                // If Gradle is run from somewhere other than the root, the task location gets trickier to translate
+                // into the projects to use; this could be implemented in the future if there's demand
+                || !rootProject.getGradle().getStartParameter().getCurrentDir().equals(rootProject.getRootDir())) {
+            // Just configure every project
+            rootProject.getSubprojects().forEach(subproject -> rootProject.evaluationDependsOn(subproject.getPath()));
+            return;
+        }
+
+        Set<String> projectPathsToEval = new LinkedHashSet<>();
+        for (String taskPath : rootProject.getGradle().getStartParameter().getTaskNames()) {
+            if (!taskPath.contains(":")) {
+                // This is a task to be run in each project that defines it, e.g. "build". This should cause every
+                // project to be defined in a configuration-on-demand build.
+                rootProject
+                        .getSubprojects()
+                        .forEach(subproject -> rootProject.evaluationDependsOn(subproject.getPath()));
+                return;
+            }
+            String projectPath = taskPath.substring(0, taskPath.lastIndexOf(':'));
+            if (!projectPath.startsWith(":")) {
+                projectPath = ":" + projectPath;
+            }
+            projectPathsToEval.add(projectPath);
+        }
+
+        Set<String> alreadyVisited = new LinkedHashSet<>();
+        while (!projectPathsToEval.isEmpty()) {
+            String projectPath = projectPathsToEval.iterator().next();
+            projectPathsToEval.remove(projectPath);
+            if (alreadyVisited.contains(projectPath) || projectPath.equals(":")) {
+                continue;
+            }
+
+            rootProject.evaluationDependsOn(projectPath);
+            Project project = rootProject.project(projectPath);
+            if (!project.getState().getExecuted()) {
+                throw new IllegalStateException(
+                        "The project has not yet been evaluated when we expect it to have been");
+            }
+            // Per
+            // https://docs.gradle.org/current/userguide/multi_project_configuration_and_execution.html#sec:configuration_on_demand,
+            // configuration is propagated transitively in two ways: dependencies and string-based task dependencies.
```

Comment:
If I'm reading this right, the docs seem to claim to support evaluating the projects that are used in task dependencies:

> The standard project dependencies are supported and makes relevant projects configured. If project A has a compile dependency on project B then building A causes configuration of both projects.

Yet it looks like we have similar logic in the code below? What is this doing?

Reply:
We can't just allow Gradle to evaluate the set of projects "naturally" because it will end up too late in the build process to do all the things we want to do afterwards, so this is imitating that logic to force early evaluation of the correct set of projects.